core/src/main/java/com/linecorp/armeria/internal/common/BodySubscriber.java

```diff
@@ -0,0 +1,306 @@
+        }
+    }
+
+    private void request0(long n) {
+        final Subscription upstream = this.upstream;
+        if (upstream == null) {
+            pendingRequests = LongMath.saturatedAdd(n, pendingRequests);
+        } else {
+            upstream.request(n);
+        }
+    }
+
+    @Override
+    public final void cancel() {
+        if (cancelCalled) {
+            return;
+        }
+        cancelCalled = true;
+        if (!notifyCancellation) {
+            downstream = NoopSubscriber.get();
+        }
+        completeTrailers(EMPTY_TRAILERS);
+        final Subscription upstream = this.upstream;
+        if (upstream != null) {
+            upstream.cancel();
+        }
+    }
+
+    @Override
+    public void onNext(HttpObject httpObject) {
+        if (httpObject instanceof HttpHeaders) {
+            final HttpHeaders trailers = (HttpHeaders) httpObject;
+            completeTrailers(trailers);
+            return;
+        }
+
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        assert downstream != null;
+        assert httpObject instanceof HttpData;
+
+        final EventExecutor executor = this.executor;
+        if (executor.inEventLoop()) {
+            onNext0((HttpData) httpObject);
+        } else {
+            executor.execute(() -> onNext0((HttpData) httpObject));
+        }
+    }
+
+    private void onNext0(HttpData httpData) {
+        wroteAny = true;
+        if (!usePooledObject) {
+            httpData = PooledObjects.copyAndClose(httpData);
+        }
+        downstream.onNext(httpData);
+    }
+
+    @Override
+    public void onComplete() {
+        final EventExecutor executor = this.executor;
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        if (executor == null || downstream == null) {
+            completing = true;
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            onComplete0(downstream);
+        } else {
+            executor.execute(() -> onComplete0(downstream));
+        }
+    }
+
+    private void onComplete0(Subscriber<? super HttpData> downstream) {
+        downstream.onComplete();
+    }
+
+    @Override
+    public void onError(Throwable cause) {
+        final EventExecutor executor = this.executor;
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        if (executor == null || downstream == null) {
+            this.cause = cause;
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            onError0(cause, downstream);
+        } else {
+            executor.execute(() -> onError0(cause, downstream));
+        }
+    }
+
+    private void onError0(Throwable cause, Subscriber<? super HttpData> downstream) {
+        downstream.onError(cause);
+        this.downstream = NoopSubscriber.get();
+    }
+
+    /**
+     * Completes the specified trailers.
+     */
+    protected final void completeTrailers(HttpHeaders trailers) {
```

Comment:
We can make this private because `DefaultSplitHttpRequest` does not have to call this.
`onComplete` and `onError` will call this method.